tests/tree_inference/test_simulate.py

```diff
@@ -264,3 +265,174 @@ def test_false_homo_mutated(
+    sp_matrix = sim.floyd_warshall(A)
+
+    assert np.array_equal(sp_matrix, SP)
+
+
+def test_shortest_path_to_ancestry_matrix_SCITE_example():
+    """Manual test of Floyd Warshall"""
+    SP = np.array([[1, -1, -1, -1], [-1, 1, 1, -1], [-1, -1, 1, -1], [1, 1, 2, 1]])
+    ancestor_test_mat = sim.shortest_path_to_ancestry_matrix(SP)
+    ancestor_true_mat = np.array(
+        [[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 1, 0], [1, 1, 1, 1]]
+    )
+
+    assert np.array_equal(ancestor_test_mat, ancestor_true_mat)
+
+
+def test_built_perfect_mutation_matrix():
+    """Manual test of building mutation matrix from ancestor matrix
+    and cell placement vector sigma by Eqn 11. - SCITE paper example"""
+    ancestor_matrix = np.array([[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 1, 0], [1, 1, 1, 1]])
+    sigma = np.array([1, 1, 1, 4, 3, 3, 2])
+
+    mutation_matrix = sim.built_perfect_mutation_matrix(4, ancestor_matrix, sigma)
+    mutation_matrix_true = np.array(
+        [
+            [1, 1, 1, 0, 0, 0, 0],
+            [0, 0, 0, 0, 1, 1, 1],
+            [0, 0, 0, 0, 1, 1, 0],
+            [1, 1, 1, 1, 1, 1, 1],
+        ]
+    )
+
+    assert np.array_equal(mutation_matrix, mutation_matrix_true)
+
+
+@pytest.mark.parametrize("seed,", [42])
+@pytest.mark.parametrize("n_nodes,", [3, 10])
+@pytest.mark.parametrize("n_cells", [10, 20])
+@pytest.mark.parametrize(
+    "strategy",
+    [
+        sim.CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT,
+        sim.CellAttachmentStrategy.UNIFORM_EXCLUDE_ROOT,
+    ],
+)
+def test_attach_cells_to_tree_for_strategy_check_bool(
+    n_nodes: int, seed: int, strategy: sim.CellAttachmentStrategy, n_cells: int
+):
+    """Test of cell attachment strategy was respected."""
+
+    rng = random.PRNGKey(seed)
+    tree = random.choice(rng, 2, shape=(n_nodes, n_nodes))
+
+    # nodes need to be their own parent in the SCITE implementation
+    diag_indices = jnp.diag_indices(tree.shape[0])
+    tree = tree.at[diag_indices].set(1)
+
+    mutation_matrix = sim.attach_cells_to_tree(rng, tree, n_cells, strategy)
+
+    if sim.CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT == strategy:
+        pass
+    elif sim.CellAttachmentStrategy.UNIFORM_EXCLUDE_ROOT == strategy:
+        pass
+    else:
+        raise TypeError(
+            "CellAttachmentStrategy not known, dimensions of mutation matrix may fail."
+        )
+
+    # check is a boolean matrix
+    assert np.array_equal(mutation_matrix, mutation_matrix.astype(bool))
+    # check for dimensions
+    assert mutation_matrix.shape == (n_nodes, n_cells)
+
+
+@pytest.mark.parametrize("seed,", [32])
+@pytest.mark.parametrize("n_nodes,", [4])
+@pytest.mark.parametrize("n_cells", [5])
+@pytest.mark.parametrize(
+    "strategy",
+    [
+        sim.CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT,
+    ],
+)
+def test_attach_cells_to_tree_case1(
+    n_nodes: int, seed: int, strategy: sim.CellAttachmentStrategy, n_cells: int
+):
+    """Manual test of attach cells to tree."""
```

Comment:
I only did one manual check of the final attach_cells_to_tree() function, yet tested all subfunctions extensively. Could not think of a good way to test for correct mutation matrices, apart from the manual tests, shape and dimensions. I am quite confident is works.

Reply:
Cool, let's proceed, then! :slightly_smiling_face: